From 3f91b2d7460fb69438ba7563635558566c86d578 Mon Sep 17 00:00:00 2001
From: Muna Sinada <quic_msinada@quicinc.com>
Date: Wed, 19 Oct 2022 13:19:59 -0700
Subject: [PATCH] hostapd: readd check for hw EHT support in EHT Operation IE

Currently EHT Operation IE is present when hardware does not support
EHT due to regression from package upgrade. Readd check for hardware
EHT support before adding EHT Operation IE.

Signed-off-by: Muna Sinada <quic_msinada@quicinc.com>
---
 src/ap/ieee802_11_eht.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -210,6 +210,7 @@ u8 * hostapd_eid_eht_capab(struct hostap
 u8 * hostapd_eid_eht_operation(struct hostapd_data *hapd, u8 *eid,
 			       enum ieee80211_op_mode opmode)
 {
+	struct hostapd_hw_modes *mode;
 	struct hostapd_config *conf = hapd->iconf;
 	struct ieee80211_eht_operation *oper;
 	u8 *pos = eid, seg0 = 0, seg1 = 0, *length_pos;
@@ -217,7 +218,10 @@ u8 * hostapd_eid_eht_operation(struct ho
 	enum oper_chan_width chwidth;
 	size_t elen = 1 + 4 + 3;
 
-	if (!hapd->iface->current_mode)
+	mode = hapd->iface->current_mode;
+	if (!mode)
+		return eid;
+	if (!mode->eht_capab[opmode].eht_supported)
 		return eid;
 
 	*pos++ = WLAN_EID_EXTENSION;
@@ -226,6 +230,7 @@ u8 * hostapd_eid_eht_operation(struct ho
 
 	oper = (struct ieee80211_eht_operation *) pos;
 	oper->oper_params = EHT_OPER_INFO_PRESENT;
+	eht_cap = &mode->eht_capab[opmode];
 
 	eht_cap = &hapd->iface->current_mode->eht_capab[opmode];
 	memcpy(&oper->basic_mcs_nss, &eht_cap->mcs,
